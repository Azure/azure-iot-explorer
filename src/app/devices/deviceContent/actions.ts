/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import actionCreatorFactory from 'typescript-fsa';
import * as actionPrefixes from '../../constants/actionPrefixes';
import * as actionTypes from '../../constants/actionTypes';
import { ModelDefinition } from '../../api/models/modelDefinition';
import { Twin } from '../../api/models/device';
import { DeviceIdentity } from '../../api/models/deviceIdentity';
import { DigitalTwinInterfaces } from './../../api/models/digitalTwinModels';
import { REPOSITORY_LOCATION_TYPE } from './../../constants/repositoryLocationTypes';

const deviceContentCreator = actionCreatorFactory(actionPrefixes.DEVICECONTENT);
const clearModelDefinitionsAction = deviceContentCreator(actionTypes.CLEAR_MODEL_DEFINITIONS);
const cloudToDeviceMessageAction = deviceContentCreator.async<CloudToDeviceMessageActionParameters, string>(actionTypes.CLOUD_TO_DEVICE_MESSAGE);
const getDeviceIdentityAction = deviceContentCreator.async<string, DeviceIdentity> (actionTypes.GET_DEVICE_IDENTITY);
const getDigitalTwinInterfacePropertiesAction = deviceContentCreator.async<string, DigitalTwinInterfaces>(actionTypes.GET_DIGITAL_TWIN_INTERFACE_PROPERTIES);
const getTwinAction = deviceContentCreator.async<string, Twin>(actionTypes.GET_TWIN);
const getModelDefinitionAction = deviceContentCreator.async<GetModelDefinitionActionParameters, ModelDefinitionActionResult>(actionTypes.FETCH_MODEL_DEFINITION);
const invokeDirectMethodAction = deviceContentCreator.async<InvokeMethodActionParameters, string>(actionTypes.INVOKE_DEVICE_METHOD);
const invokeDigitalTwinInterfaceCommandAction = deviceContentCreator.async<InvokeDigitalTwinInterfaceCommandActionParameters, string>(actionTypes.INVOKE_DIGITAL_TWIN_INTERFACE_COMMAND);
const patchDigitalTwinInterfacePropertiesAction = deviceContentCreator.async<PatchDigitalTwinInterfacePropertiesActionParameters, DigitalTwinInterfaces>(actionTypes.PATCH_DIGITAL_TWIN_INTERFACE_PROPERTIES);
const setInterfaceIdAction = deviceContentCreator<string>(actionTypes.SET_INTERFACE_ID);
const updateDeviceIdentityAction = deviceContentCreator.async<DeviceIdentity, DeviceIdentity> (actionTypes.UPDATE_DEVICE_IDENTITY);
const updateTwinAction = deviceContentCreator.async<UpdateTwinActionParameters, Twin>(actionTypes.UPDATE_TWIN);
export {
    clearModelDefinitionsAction,
    cloudToDeviceMessageAction,
    getDeviceIdentityAction,
    getDigitalTwinInterfacePropertiesAction,
    getTwinAction,
    getModelDefinitionAction,
    invokeDirectMethodAction,
    invokeDigitalTwinInterfaceCommandAction,
    patchDigitalTwinInterfacePropertiesAction,
    setInterfaceIdAction,
    updateTwinAction,
    updateDeviceIdentityAction
};

export interface CloudToDeviceMessageActionParameters {
    deviceId: string;
    body: string;
    properties?: Array<{key: string, value: string, isSystemProperty: boolean}>;
}

export interface PatchDigitalTwinInterfacePropertiesActionParameters {
    digitalTwinId: string;
    interfacesPatchData: any; // tslint:disable-line:no-any
    propertyKey: string;
}

export interface InvokeDigitalTwinInterfaceCommandActionParameters {
    digitalTwinId: string;
    commandName: string;
    commandPayload: any; // tslint:disable-line:no-any
    propertyKey?: string;
}

export interface InvokeMethodActionParameters {
    connectTimeoutInSeconds: number;
    deviceId: string;
    methodName: string;
    payload?: object;
    responseTimeoutInSeconds: number;
}

export interface UpdateTwinActionParameters {
    deviceId: string;
    twin: Twin;
}

export interface ModelDefinitionActionResult {
    modelDefinition: ModelDefinition;
    source: REPOSITORY_LOCATION_TYPE;
}

export interface GetModelDefinitionActionParameters {
    digitalTwinId: string;
    interfaceId: string;
}
