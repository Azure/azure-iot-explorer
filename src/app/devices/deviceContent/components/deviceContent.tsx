/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { Route, RouteComponentProps } from 'react-router-dom';
import { IconButton } from 'office-ui-fabric-react/lib/Button';
import DeviceIdentityContainer from './deviceIdentity/deviceIdentityContainer';
import DeviceTwinContainer from './deviceTwin/deviceTwinContainer';
import DeviceEventsContainer from './deviceEvents/deviceEventsContainer';
import DirectMethodContainer from './directMethod/directMethodContainer';
import CloudToDeviceMessageContainer from './cloudToDeviceMessage/cloudToDeviceMessageContainer';
import ModuleIdentityContent from '../../module/components/moduleIdentity/moduleIdentityContent';
import DeviceContentNavComponent from './deviceContentNav';
import Breadcrumb from '../../../shared/components/breadcrumb';
import DigitalTwinsContentContainer from './digitalTwin/digitalTwinContentContainer';
import { DigitalTwinInterfacesContainer } from './digitalTwin/digitalTwinInterfaces';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { LocalizationContextConsumer, LocalizationContextInterface } from '../../../shared/contexts/localizationContext';
import { NAV } from '../../../constants/iconNames';
import { ROUTE_PARTS, ROUTE_PARAMS } from '../../../constants/routes';
import { DeviceIdentity } from '../../../api/models/deviceIdentity';
import { SynchronizationWrapper } from '../../../api/models/synchronizationWrapper';
import { SynchronizationStatus } from '../../../api/models/synchronizationStatus';
import MultiLineShimmer from '../../../shared/components/multiLineShimmer';
import { getDeviceIdFromQueryString } from '../../../shared/utils/queryStringHelper';
import '../../../css/_deviceContent.scss';
import '../../../css/_layouts.scss';

interface DeviceContentState {
    appMenuVisible: boolean;
}
export interface DeviceContentDataProps {
    deviceId: string;
    interfaceIds: string[];
    isLoading: boolean;
    isPnPDevice: boolean;
    identityWrapper: SynchronizationWrapper<DeviceIdentity>;
}

export interface DeviceContentDispatchProps {
    setComponentName: (componentName: string) => void;
    getDigitalTwinInterfaceProperties: (deviceId: string) => void;
    getDeviceIdentity: (deviceId: string) => void;
}

export type DeviceContentProps = DeviceContentDataProps & DeviceContentDispatchProps & RouteComponentProps;

export class DeviceContentComponent extends React.PureComponent<DeviceContentProps, DeviceContentState> {
    constructor(props: DeviceContentProps) {
        super(props);
        this.state = {
            appMenuVisible: true
        };
    }

    public render(): JSX.Element {
        return (
            <LocalizationContextConsumer>
            {(context: LocalizationContextInterface) => (
                <div>
                    {this.props.deviceId &&
                        <div className="edit-content">
                            <div className="device-content">
                                <>
                                    {this.renderNav(context)}
                                    {this.renderDeviceContentDetail()}
                                </>
                            </div>
                        </div>
                    }
                </div>
            )}
            </LocalizationContextConsumer>
        );
    }

    public componentDidMount() {
        this.props.getDigitalTwinInterfaceProperties(this.props.deviceId);
        this.props.getDeviceIdentity(this.props.deviceId);
    }

    public componentDidUpdate(oldProps: DeviceContentProps) {
        if (getDeviceIdFromQueryString(oldProps) !== getDeviceIdFromQueryString(this.props)) {
            const deviceId = getDeviceIdFromQueryString(this.props);
            this.props.getDeviceIdentity(deviceId);
            this.props.getDigitalTwinInterfaceProperties(deviceId);
            const url = this.props.match.url;
            this.props.history.push(`${url}/${ROUTE_PARTS.IDENTITY}/?${ROUTE_PARAMS.DEVICE_ID}=${encodeURIComponent(deviceId)}`);
        }
    }

    private readonly renderNav = (context: LocalizationContextInterface) => {
        return (
            <div className={'device-content-nav-bar' + (!this.state.appMenuVisible ? ' collapsed' : '')}>
                <nav>
                    <div className="navToggle">
                        <IconButton
                            tabIndex={0}
                            iconProps={{ iconName: NAV }}
                            title={this.state.appMenuVisible ? context.t(ResourceKeys.deviceContent.navBar.collapse) : context.t(ResourceKeys.deviceContent.navBar.expand)}
                            ariaLabel={this.state.appMenuVisible ? context.t(ResourceKeys.deviceContent.navBar.collapse) : context.t(ResourceKeys.deviceContent.navBar.expand)}
                            onClick={this.collapseToggle}
                        />
                    </div>
                    <div className="nav-links">
                        {this.createNavLinks()}
                    </div>
                </nav>
            </div>
        );
    }

    private readonly renderDeviceContentDetail = () => {
        const url = this.props.match.url;

        return (
            <div className={'device-content-detail' + (!this.state.appMenuVisible ? ' collapsed' : '')}>
                <Route path={`${url}/${ROUTE_PARTS.IDENTITY}/`} component={DeviceIdentityContainer} />
                <Route path={`${url}/${ROUTE_PARTS.TWIN}/`} component={DeviceTwinContainer} />
                <Route path={`${url}/${ROUTE_PARTS.EVENTS}/`} component={DeviceEventsContainer}/>
                <Route path={`${url}/${ROUTE_PARTS.METHODS}/`} component={DirectMethodContainer} />
                <Route path={`${url}/${ROUTE_PARTS.CLOUD_TO_DEVICE_MESSAGE}/`} component={CloudToDeviceMessageContainer} />
                <Route path={`${url}/${ROUTE_PARTS.MODULE_IDENTITY}/`} component={ModuleIdentityContent} />
                <Route exact={true} path={`${url}/${ROUTE_PARTS.DIGITAL_TWINS}/`} component={DigitalTwinInterfacesContainer} />
                <Route path={`${url}/${ROUTE_PARTS.DIGITAL_TWINS}/${ROUTE_PARTS.DIGITAL_TWINS_DETAIL}/`} component={DigitalTwinsContentContainer}/>
            </div>
        );
    }

    private readonly collapseToggle = () => {
        this.setState({
            appMenuVisible: !this.state.appMenuVisible
        });
    }

    private readonly createNavLinks = () => {
        return (
            this.props.identityWrapper && this.props.identityWrapper.synchronizationStatus === SynchronizationStatus.working ?
                <MultiLineShimmer/> :
                (
                    <DeviceContentNavComponent
                        {...this.props}
                        isEdgeDevice={this.props.identityWrapper && this.props.identityWrapper.payload && this.props.identityWrapper.payload.capabilities.iotEdge}
                    />
                )
        );
    }
}
