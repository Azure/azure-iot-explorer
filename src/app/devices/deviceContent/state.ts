/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { Record } from 'immutable';
import { IM } from '../../shared/types/types';
import { ModelDefinitionWithSourceWrapper } from '../../api/models/modelDefinitionWithSourceWrapper';
import { DeviceIdentityWrapper } from '../../api/models/deviceIdentityWrapper';
import { DeviceTwinWrapper } from './../../api/models/deviceTwinWrapper';
import { DigitalTwinInterfacePropertiesWrapper } from '../../api/models/digitalTwinInterfacePropertiesWrapper';

export interface DeviceContentStateInterface {
    deviceIdentity: DeviceIdentityWrapper;
    deviceTwin: DeviceTwinWrapper;
    digitalTwinInterfaceProperties: DigitalTwinInterfacePropertiesWrapper;
    interfaceIdSelected: string;
    modelDefinitionWithSource: ModelDefinitionWithSourceWrapper;
}

export const deviceContentStateInitial = Record<DeviceContentStateInterface>({
    deviceIdentity: null,
    deviceTwin: null,
    digitalTwinInterfaceProperties: null,
    interfaceIdSelected: '',
    modelDefinitionWithSource: null,
});

export type DeviceContentStateType = IM<DeviceContentStateInterface>;
