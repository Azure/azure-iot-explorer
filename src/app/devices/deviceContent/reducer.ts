/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { reducerWithInitialState } from 'typescript-fsa-reducers';
import { DeviceContentStateType, deviceContentStateInitial } from './state';
import {
    clearModelDefinitionsAction,
    getDeviceIdentityAction,
    getDigitalTwinInterfacePropertiesAction,
    getTwinAction,
    getModelDefinitionAction,
    setInterfaceIdAction,
    updateTwinAction,
    UpdateTwinActionParameters,
    updateDeviceIdentityAction,
    patchDigitalTwinInterfacePropertiesAction,
    PatchDigitalTwinInterfacePropertiesActionParameters,
    ModelDefinitionActionResult,
    GetModelDefinitionActionParameters,
    getModuleIdentitiesAction,
    addModuleIdentityAction
} from './actions';
import { Twin } from '../../api/models/device';
import { DeviceIdentity } from '../../api/models/deviceIdentity';
import { SynchronizationStatus } from '../../api/models/synchronizationStatus';
import { DigitalTwinInterfaces } from '../../api/models/digitalTwinModels';
import { ModuleIdentity } from './../../api/models/moduleIdentity';

const reducer = reducerWithInitialState<DeviceContentStateType>(deviceContentStateInitial())
    //#region DeviceIdentity-related actions
    .case(getDeviceIdentityAction.started, (state: DeviceContentStateType) => {
        return state.merge({
            deviceIdentity: {
                deviceIdentitySynchronizationStatus: SynchronizationStatus.working
            }
        });
    })
    .case(getDeviceIdentityAction.done, (state: DeviceContentStateType, payload: {params: string} & {result: DeviceIdentity}) => {
        return state.merge({
            deviceIdentity: {
                deviceIdentity: payload.result,
                deviceIdentitySynchronizationStatus: SynchronizationStatus.fetched
            }
        });
    })
    .case(getDeviceIdentityAction.failed, (state: DeviceContentStateType) => {
        return state.merge({
            deviceIdentity: {
                deviceIdentitySynchronizationStatus: SynchronizationStatus.failed
            }
        });
    })
    .case(updateDeviceIdentityAction.started, (state: DeviceContentStateType) => {
        return state.merge({
            deviceIdentity: {
                deviceIdentity: state.deviceIdentity.deviceIdentity,
                deviceIdentitySynchronizationStatus: SynchronizationStatus.updating
            }
        });
    })
    .case(updateDeviceIdentityAction.done, (state: DeviceContentStateType, payload: {params: DeviceIdentity} & {result: DeviceIdentity}) => {
        return state.merge({
            deviceIdentity: {
                deviceIdentity: payload.result,
                deviceIdentitySynchronizationStatus: SynchronizationStatus.upserted
            }
        });
    })
    .case(updateDeviceIdentityAction.failed, (state: DeviceContentStateType) => {
        return state.merge({
            deviceIdentity: {
                deviceIdentity: state.deviceIdentity.deviceIdentity,
                deviceIdentitySynchronizationStatus: SynchronizationStatus.failed
            }
        });
    })
    //#endregion
    //#region DeviceTwin-related actions
    .case(getTwinAction.started, (state: DeviceContentStateType) => {
        return state.merge({
            deviceTwin: {
                deviceTwinSynchronizationStatus: SynchronizationStatus.working
            }
        });
    })
    .case(getTwinAction.done, (state: DeviceContentStateType, payload: {params: string} & {result: Twin}) => {
        return state.merge({
            deviceTwin: {
                deviceTwin: payload.result,
                deviceTwinSynchronizationStatus: SynchronizationStatus.fetched
            }
        });
    })
    .case(getTwinAction.failed, (state: DeviceContentStateType) => {
        return state.merge({
            deviceTwin: {
                deviceTwinSynchronizationStatus: SynchronizationStatus.failed
            }
        });
    })
    .case(updateTwinAction.started, (state: DeviceContentStateType) => {
        return state.merge({
            deviceTwin: {
                deviceTwin: state.deviceTwin.deviceTwin,
                deviceTwinSynchronizationStatus: SynchronizationStatus.updating
            }
        });
    })
    .case(updateTwinAction.done, (state: DeviceContentStateType, payload: {params: UpdateTwinActionParameters} & {result: Twin}) => {
        return state.merge({
            deviceTwin: {
                deviceTwin: payload.result,
                deviceTwinSynchronizationStatus: SynchronizationStatus.upserted
            }
        });
    })
    .case(updateTwinAction.failed, (state: DeviceContentStateType) => {
        return state.merge({
            deviceTwin: {
                deviceTwin: state.deviceTwin.deviceTwin,
                deviceTwinSynchronizationStatus: SynchronizationStatus.failed
            }
        });
    })
    //#endregion
    //#region Interface-related actions
    .case(getModelDefinitionAction.started, (state: DeviceContentStateType) => {
        return state.merge({
            modelDefinitionWithSource: {
                modelDefinitionSynchronizationStatus: SynchronizationStatus.working
            }
        });
    })
    .case(getModelDefinitionAction.done, (state: DeviceContentStateType, payload: {params: GetModelDefinitionActionParameters} & {result: ModelDefinitionActionResult}) => {
        return state.merge({
            modelDefinitionWithSource: {
                modelDefinition: payload.result.modelDefinition,
                modelDefinitionSynchronizationStatus: SynchronizationStatus.fetched,
                source: payload.result.source
            }
        });
    })
    .case(getModelDefinitionAction.failed, (state: DeviceContentStateType) => {
        return state.merge({
            modelDefinitionWithSource: {
                modelDefinition: null,
                modelDefinitionSynchronizationStatus: SynchronizationStatus.failed
            }
        });
    })
    .case(clearModelDefinitionsAction, (state: DeviceContentStateType) => {
        return state.merge({
            modelDefinitionWithSource: null
        });
    })
    .case(setInterfaceIdAction, (state: DeviceContentStateType, payload: string) => {
        return state.merge({
            interfaceIdSelected: payload
        });
    })
    //#endregion
    //#region DigitalTwin-related actions
    .case(getDigitalTwinInterfacePropertiesAction.started, (state: DeviceContentStateType) => {
        return state.merge({
            digitalTwinInterfaceProperties: {
                digitalTwinInterfacePropertiesSyncStatus: SynchronizationStatus.working
            }
        });
    })
    .case(getDigitalTwinInterfacePropertiesAction.done, (state: DeviceContentStateType, payload: {params: string} & {result: DigitalTwinInterfaces}) => {
        return state.merge({
            digitalTwinInterfaceProperties: {
                digitalTwinInterfaceProperties: payload.result,
                digitalTwinInterfacePropertiesSyncStatus: SynchronizationStatus.fetched
            }
        });
    })
    .case(getDigitalTwinInterfacePropertiesAction.failed, (state: DeviceContentStateType) => {
        return state.merge({
            digitalTwinInterfaceProperties: {
                digitalTwinInterfacePropertiesSyncStatus: SynchronizationStatus.failed
            }
        });
    })
    .case(patchDigitalTwinInterfacePropertiesAction.started, (state: DeviceContentStateType) => {
        return state.merge({
            digitalTwinInterfaceProperties: {
                digitalTwinInterfaceProperties: state.digitalTwinInterfaceProperties.digitalTwinInterfaceProperties,
                digitalTwinInterfacePropertiesSyncStatus: SynchronizationStatus.updating
            }
        });
    })
    .case(patchDigitalTwinInterfacePropertiesAction.done, (state: DeviceContentStateType, payload: {params: PatchDigitalTwinInterfacePropertiesActionParameters} & {result: DigitalTwinInterfaces}) => {
        return state.merge({
            digitalTwinInterfaceProperties: {
                digitalTwinInterfaceProperties: payload.result,
                digitalTwinInterfacePropertiesSyncStatus: SynchronizationStatus.upserted
            }
        });
    })
    .case(patchDigitalTwinInterfacePropertiesAction.failed, (state: DeviceContentStateType) => {
        return state.merge({
            digitalTwinInterfaceProperties: {
                digitalTwinInterfaceProperties: state.digitalTwinInterfaceProperties.digitalTwinInterfaceProperties,
                digitalTwinInterfacePropertiesSyncStatus: SynchronizationStatus.failed
            }
        });
    })
    //#endregion
    //#region ModuleIdentity-related actions
    .case(getModuleIdentitiesAction.started, (state: DeviceContentStateType) => {
        return state.merge({
            moduleIdentityList: {
                synchronizationStatus: SynchronizationStatus.working
            }
        });
    })
    .case(getModuleIdentitiesAction.done, (state: DeviceContentStateType, payload: {params: string} & {result: ModuleIdentity[]}) => {
        return state.merge({
            moduleIdentityList: {
                moduleIdentities: payload.result,
                synchronizationStatus: SynchronizationStatus.fetched
            }
        });
    })
    .case(getModuleIdentitiesAction.failed, (state: DeviceContentStateType) => {
        return state.merge({
            moduleIdentityList: {
                synchronizationStatus: SynchronizationStatus.failed
            }
        });
    })
    .case(addModuleIdentityAction.started, (state: DeviceContentStateType) => {
        return state.merge({
            moduleIdentityList: {
                synchronizationStatus: SynchronizationStatus.working
            }
        });
    })
    .case(addModuleIdentityAction.done, (state: DeviceContentStateType) => {
        return state.merge({
            moduleIdentityList: {
                synchronizationStatus: SynchronizationStatus.upserted
            }
        });
    })
    .case(addModuleIdentityAction.failed, (state: DeviceContentStateType) => {
        return state.merge({
            moduleIdentityList: {
                synchronizationStatus: SynchronizationStatus.failed
            }
        });
    });
    //#endregion
export default reducer;
