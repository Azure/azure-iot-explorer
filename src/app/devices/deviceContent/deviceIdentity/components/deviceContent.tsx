/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { Route, useLocation } from 'react-router-dom';
import { IconButton } from 'office-ui-fabric-react/lib/Button';
import { DeviceIdentityInformation } from './deviceIdentity';
import { DeviceTwin } from '../../deviceTwin/components/deviceTwin';
import { DeviceEvents } from '../../deviceEvents/components/deviceEvents';
import { DirectMethod } from '../../directMethod/components/directMethod';
import { CloudToDeviceMessage } from '../../cloudToDeviceMessage/components/cloudToDeviceMessage';
import { DeviceContentNavComponent } from './deviceContentNav';
import { DigitalTwinContent } from '../../components/digitalTwin/digitalTwinContent';
import { DigitalTwinInterfacesContainer } from '../../components/digitalTwin/digitalTwinInterfaces';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { useLocalizationContext } from '../../../../shared/contexts/localizationContext';
import { NAV } from '../../../../constants/iconNames';
import { ROUTE_PARTS } from '../../../../constants/routes';
import { SynchronizationStatus } from '../../../../api/models/synchronizationStatus';
import MultiLineShimmer from '../../../../shared/components/multiLineShimmer';
import { getDeviceIdFromQueryString } from '../../../../shared/utils/queryStringHelper';
import { ModuleIdentityTwin } from '../../../module/moduleIdentityTwin/components/moduleIdentityTwin';
import { AddModuleIdentity } from '../../../module/addModuleIdentity/components/addModuleIdentity';
import { ModuleIdentityList } from '../../../module/moduleIdentityList/components/moduleIdentityList';
import { ModuleIdentityDetail } from '../../../module/moduleIndentityDetail/components/moduleIdentityDetail';
import { useAsyncSagaReducer } from '../../../../shared/hooks/useAsyncSagaReducer';
import { deviceIdentityReducer } from '../reducer';
import { DeviceIdentitySaga } from '../saga';
import { deviceIdentityStateInitial } from '../state';
import { getDeviceIdentityAction, updateDeviceIdentityAction } from '../actions';
import { DeviceIdentity } from '../../../../api/models/deviceIdentity';
import '../../../../css/_deviceContent.scss';
import '../../../../css/_layouts.scss';

export const DeviceContent: React.FC = () => {
    const { t } = useLocalizationContext();
    const { search } = useLocation();
    const deviceId = getDeviceIdFromQueryString(search);

    const [ localState, dispatch ] = useAsyncSagaReducer(deviceIdentityReducer, DeviceIdentitySaga, deviceIdentityStateInitial());
    const synchronizationStatus = localState.synchronizationStatus;
    const deviceIdentity = localState.payload;

    const [ appMenuVisible, setAppMenuVisible ] = React.useState(true);

    React.useEffect(() => {
        dispatch(getDeviceIdentityAction.started(deviceId));
    },              [deviceId]);

    const renderNav = () => {
        return (
            <div className={'device-content-nav-bar' + (!appMenuVisible ? ' collapsed' : '')}>
                <nav>
                    <div className="navToggle">
                        <IconButton
                            tabIndex={0}
                            iconProps={{ iconName: NAV }}
                            title={appMenuVisible ? t(ResourceKeys.deviceContent.navBar.collapse) : t(ResourceKeys.deviceContent.navBar.expand)}
                            ariaLabel={appMenuVisible ? t(ResourceKeys.deviceContent.navBar.collapse) : t(ResourceKeys.deviceContent.navBar.expand)}
                            onClick={collapseToggle}
                        />
                    </div>
                    <div className="nav-links">
                        {createNavLinks()}
                    </div>
                </nav>
            </div>
        );
    };

    const updateDeviceIdentity = (deviceIdentityToUpdate: DeviceIdentity) => {
        dispatch(updateDeviceIdentityAction.started(deviceIdentityToUpdate));
    };

    const createDeviceIdentityComponent = () => {
        return (
            <DeviceIdentityInformation
                deviceIdentity={deviceIdentity}
                synchronizationStatus={synchronizationStatus}
                updateDeviceIdentity={updateDeviceIdentity}
            />);
    };

    const renderDeviceContentDetail = () => {
        const currentRoutePath = `/${ROUTE_PARTS.RESOURCE}/:hostName/${ROUTE_PARTS.DEVICES}/${ROUTE_PARTS.DEVICE_DETAIL}`;
        return (
            <div className={'device-content-detail' + (!appMenuVisible ? ' collapsed' : '')}>
                <Route path={`${currentRoutePath}/${ROUTE_PARTS.IDENTITY}`} component={createDeviceIdentityComponent} />
                <Route path={`${currentRoutePath}/${ROUTE_PARTS.TWIN}`} component={DeviceTwin} />
                <Route path={`${currentRoutePath}/${ROUTE_PARTS.EVENTS}`} component={DeviceEvents}/>
                <Route path={`${currentRoutePath}/${ROUTE_PARTS.METHODS}`} component={DirectMethod} />
                <Route path={`${currentRoutePath}/${ROUTE_PARTS.CLOUD_TO_DEVICE_MESSAGE}`} component={CloudToDeviceMessage} />
                <Route exact={true} path={`${currentRoutePath}/${ROUTE_PARTS.DIGITAL_TWINS}`} component={DigitalTwinInterfacesContainer} />
                <Route path={`${currentRoutePath}/${ROUTE_PARTS.DIGITAL_TWINS}/${ROUTE_PARTS.DIGITAL_TWINS_DETAIL}`} component={DigitalTwinContent}/>
                {/* module Routes */}
                <Route exact={true} path={`${currentRoutePath}/${ROUTE_PARTS.MODULE_IDENTITY}`} component={ModuleIdentityList}/>
                <Route exact={true} path={`${currentRoutePath}/${ROUTE_PARTS.MODULE_IDENTITY}/${ROUTE_PARTS.ADD}`} component={AddModuleIdentity}/>
                <Route exact={true} path={`${currentRoutePath}/${ROUTE_PARTS.MODULE_IDENTITY}/${ROUTE_PARTS.MODULE_DETAIL}`} component={ModuleIdentityDetail}/>
                <Route exact={true} path={`${currentRoutePath}/${ROUTE_PARTS.MODULE_IDENTITY}/${ROUTE_PARTS.MODULE_TWIN}`} component={ModuleIdentityTwin}/>
            </div>
        );
    };

    const collapseToggle = () => {
        setAppMenuVisible(true);
    };

    const createNavLinks = () => {
        return (
            synchronizationStatus === SynchronizationStatus.working ?
                <MultiLineShimmer/> :
                (
                    <DeviceContentNavComponent
                        isEdgeDevice={deviceIdentity && deviceIdentity.capabilities && deviceIdentity.capabilities.iotEdge}
                    />
                )
        );
    };

    return (
        <div>
            {deviceId &&
                <div className="edit-content">
                    <div className="device-content">
                        {renderNav()}
                        {renderDeviceContentDetail()}
                    </div>
                </div>
            }
        </div>
    );
};
