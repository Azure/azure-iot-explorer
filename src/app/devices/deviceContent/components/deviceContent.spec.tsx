/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { DeviceContentComponent, DeviceContentDispatchProps, DeviceContentDataProps } from './deviceContent';
import { testSnapshot } from '../../../shared/utils/testHelpers';

describe('deviceContent', () => {

    const pathname = `/#/devices/detail/?id=testDevice`;

    const location: any = { // tslint:disable-line:no-any
        pathname
    };
    const routerprops: any = { // tslint:disable-line:no-any
        history: {
            location
        },
        location,
        match: {}
    };
    const deviceContentDataProps: DeviceContentDataProps = {
        deviceId: 'testDevice',
        identityWrapper: null,
        interfaceId: 'testInterfaceId',
        interfaceIds: [],
        isLoading: false,
        isPnPDevice: true,
    };

    const deviceContentDispatchProps: DeviceContentDispatchProps = {
        getDeviceIdentity: jest.fn(),
        getDigitalTwinInterfaceProperties: jest.fn(),
        setInterfaceId: jest.fn()
    };

    const getComponent = (overrides = {}) => {
        const props = {
            ...deviceContentDataProps,
            ...deviceContentDispatchProps,
            ...routerprops,
            ...overrides
        };

        return (
            <DeviceContentComponent {...props} />
        );
    };

    it('matches snapshot', () => {
        testSnapshot(getComponent());
    });
});
