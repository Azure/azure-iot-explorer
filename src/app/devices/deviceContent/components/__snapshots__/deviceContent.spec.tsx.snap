// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deviceContent matches snapshot 1`] = `
<div>
  <div
    className="edit-content"
  >
    <div
      className="device-content"
    >
      <div
        className="device-content-nav-bar"
      >
        <nav>
          <div
            className="navToggle"
          >
            <CustomizedIconButton
              ariaLabel="deviceContent.navBar.collapse"
              iconProps={
                Object {
                  "iconName": "GlobalNavButton",
                }
              }
              onClick={[Function]}
              tabIndex={0}
              title="deviceContent.navBar.collapse"
            />
          </div>
          <div
            className="nav-links"
          >
            <DeviceContentNavComponent
              deviceId="testDevice"
              getDeviceIdentity={
                [MockFunction] {
                  "calls": Array [
                    Array [
                      "testDevice",
                    ],
                  ],
                  "results": Array [
                    Object {
                      "type": "return",
                      "value": undefined,
                    },
                  ],
                }
              }
              getDigitalTwinInterfaceProperties={
                [MockFunction] {
                  "calls": Array [
                    Array [
                      "testDevice",
                    ],
                  ],
                  "results": Array [
                    Object {
                      "type": "return",
                      "value": undefined,
                    },
                  ],
                }
              }
              history={
                Object {
                  "location": Object {
                    "pathname": "/#/devices/detail/?id=testDevice",
                  },
                }
              }
              identityWrapper={null}
              interfaceIds={Array []}
              isEdgeDevice={null}
              isLoading={false}
              isPnPDevice={true}
              location={
                Object {
                  "pathname": "/#/devices/detail/?id=testDevice",
                }
              }
              match={Object {}}
              setComponentName={[MockFunction]}
            />
          </div>
        </nav>
      </div>
      <div
        className="device-content-detail"
      >
        <Route
          component={[Function]}
          path="undefined/identity/"
        />
        <Route
          component={[Function]}
          path="undefined/twin/"
        />
        <Route
          component={
            Object {
              "$$typeof": Symbol(react.memo),
              "WrappedComponent": [Function],
              "compare": null,
              "displayName": "Connect(DeviceEventsComponent)",
              "type": [Function],
            }
          }
          path="undefined/events/"
        />
        <Route
          component={[Function]}
          path="undefined/methods/"
        />
        <Route
          component={[Function]}
          path="undefined/cloudToDeviceMessage/"
        />
        <Route
          component={[Function]}
          path="undefined/moduleIdentity/"
        />
        <Route
          component={[Function]}
          exact={true}
          path="undefined/ioTPlugAndPlay/"
        />
        <Route
          component={[Function]}
          path="undefined/ioTPlugAndPlay/ioTPlugAndPlayDetail/"
        />
      </div>
    </div>
  </div>
</div>
`;
